In [61]:
import urllib.request
import requests
import math 
from numpy import *
from numpy.linalg import norm
a = 0.4
b = 0.2
r = float(requests.get(f"http://ramcdougal.com/cgi-bin/error_function.py?a={a}&b={b}", headers={"User-Agent": "MyScript"}).text.strip())



In [62]:
def f(a, b):
    r = float(requests.get(f"http://ramcdougal.com/cgi-bin/error_function.py?a={a}&b={b}", headers={"User-Agent": "MyScript"}).text.strip())
    return r

In [63]:
f(0.4,0.2)

1.294915

In [68]:
def dfda(a,b,h):
    return (f(a+h,b)-f(a,b))/h #Based on the alagorithem 
def dfdb(a,b,h):
    return (f(a,b+h)-f(a,b))/h
def gradf(a,b,h):
    return array([dfda(a,b,h), dfdb(a,b,h)])

In [69]:
#https://www.cs.toronto.edu/~guerzhoy/411/lec/W02/python/graddescent2d.html
def grad_descent2(f, gradf, init_t, alpha): #t = initial guess, alpha = gamma(step size) 
    EPS = 1e-8 #EPS:Epsilon is a very small number to prevent any division by zero 
    #in the implementation (e.g. 10E-8).
    h = 1e-3
    prev_t = init_t-10*EPS
    t = init_t.copy()
    
    max_iter = 1000#In practice, the Maximum Number of Steps = 1,000 or greater.
    #So I set max_steps = 1000
    iter = 0
    while norm(t - prev_t) > EPS and iter < max_iter: #Stopping rule:
        #If there have been more than Maximum Number of Steps,Gradient Descent would stop.
        prev_t = t.copy()
        t -= alpha*gradf(t[0], t[1],h)
        print (t, f(t[0], t[1]), gradf(t[0], t[1],h))
        iter += 1
    
    return t

In [66]:
f(0.5,0.5)

1.216377

In [70]:
grad_descent2(f, gradf, array([0.01, 0.01]), 0.1) #step size=0.1
#so it can descent quickly in the beginning.

[0.0511 0.1457] 1.4223669 [-0.3288 -1.0856]
[0.08398 0.25426] 1.306428148 [-0.26304 -0.86848]
[0.110284 0.341108] 1.23220471632 [-0.210432 -0.694784]
[0.1313272 0.4105864] 1.18468361572 [-0.1683456 -0.5558272]
[0.14816176 0.46616912] 1.15425562789 [-0.13467648 -0.44466176]
[0.16162941 0.5106353 ] 1.13477012891 [-0.10774119 -0.35572941]
[0.17240353 0.54620824] 1.12229014004 [-0.08619295 -0.28458353]
[0.18102282 0.57466659] 1.11429553162 [-0.06895435 -0.22766682]
[0.18791826 0.59743327] 1.10917304997 [-0.05516349 -0.18213346]
[0.19343461 0.61564662] 1.10588991566 [-0.04413079 -0.14570677]
[0.19784769 0.6302173 ] 1.10378491295 [-0.03530463 -0.11656541]
[0.20137815 0.64187384] 1.10243467389 [-0.0282437  -0.09325233]
[0.20420252 0.65119907] 1.10156809097 [-0.02259497 -0.07460187]
[0.20646202 0.65865926] 1.1010115339 [-0.01807597 -0.05968148]
[0.20826961 0.6646274 ] 1.10065378233 [-0.01446077 -0.04774519]
[0.20971569 0.66940192] 1.10042357719 [-0.01156863 -0.03819616]
[0.21087255 0.67322154]

array([0.21549999, 0.68849996])

In [ ]:
#In practice, the Maximum Number of Steps = 1,000 or greater
#If there have been more than Maximum Number of Steps, 
#Gradient Descent would stop.
#step size = slop * learning rate
#New intercept = old intercept - step size


In [71]:
grad_descent2(f, gradf, array([0.001, 0.001]), 0.1)

[0.0439 0.1385] 1.43266866 [-0.3432 -1.1   ]
[0.07822 0.2485 ] 1.3130235784 [-0.27456 -0.88   ]
[0.105676 0.3365  ] 1.23642763498 [-0.219648 -0.704   ]
[0.1276408 0.4069   ] 1.18738775822 [-0.1757184 -0.5632   ]
[0.14521264 0.46322   ] 1.15598745874 [-0.14057472 -0.45056   ]
[0.15927011 0.508276  ] 1.13587944437 [-0.11245978 -0.360448  ]
[0.17051609 0.5443208 ] 1.12300085698 [-0.08996782 -0.2883584 ]
[0.17951287 0.57315664] 1.11475099457 [-0.07197426 -0.23068672]
[0.1867103  0.59622531] 1.10946502938 [-0.05757941 -0.18454938]
[0.19246824 0.61468025] 1.10607716902 [-0.04606353 -0.1476395 ]
[0.19707459 0.6294442 ] 1.10390506438 [-0.03685081 -0.1181116 ]
[0.20075967 0.64125536] 1.10251181822 [-0.02948066 -0.09448928]
[0.20370774 0.65070429] 1.10161766124 [-0.02358452 -0.07559143]
[0.20606619 0.65826343] 1.10104341724 [-0.01886762 -0.06047314]
[0.20795295 0.66431074] 1.10067431428 [-0.0150941  -0.04837851]
[0.20946236 0.6691486 ] 1.10043681894 [-0.01207528 -0.03870281]
[0.21066989 0.673018

array([0.21549999, 0.68849996])

In [72]:
grad_descent2(f, gradf, array([0.99, 0.01]), 0.1)

[0.8229 0.1051] 1.04914606 [ 0.6684 -0.3804]
[0.75606 0.14314] 1.0078300696 [ 0.26736 -0.15216]
[0.729324 0.158356] 1.00124024714 [ 0.106944 -0.060864]
[0.7186296 0.1644424] 1.00019416994 [ 0.0427776 -0.0243456]
[0.71435184 0.16687696] 1.00003011535 [ 0.01711104 -0.00973824]
[0.71264074 0.16785078] 1.00000519372 [ 0.00684442 -0.0038953 ]
[0.71195629 0.16824031] 1.0000017371 [ 0.00273776 -0.00155812]
[0.71168252 0.16839613] 1.00000139638 [ 0.0010951  -0.00062325]
[0.71157301 0.16845845] 1.00000142679 [ 0.00043805 -0.00024929]
[0.7115292  0.16848338] 1.00000146564 [ 1.7522e-04 -9.9720e-05]
[0.71151168 0.16849335] 1.00000148544 [ 7.009e-05 -3.989e-05]
[0.71150467 0.16849734] 1.00000149405 [ 2.80300001e-05 -1.59600000e-05]
[0.71150187 0.16849894] 1.0000014976 [ 1.12100000e-05 -6.38000008e-06]
[0.71150075 0.16849958] 1.00000149903 [ 4.49000015e-06 -2.54999999e-06]
[0.7115003  0.16849983] 1.00000149961 [ 1.80000015e-06 -1.01999986e-06]
[0.71150012 0.16849993] 1.00000149985 [ 7.10000059e-07 -

array([0.7115, 0.1685])

In [73]:
grad_descent2(f, gradf, array([0.01, 0.99]), 0.1)

[0.0511 0.9297] 1.1851285 [-0.3288  0.4824]
[0.08398 0.88146] 1.154470132 [-0.26304  0.38592]
[0.110284 0.842868] 1.13485123408 [-0.210432  0.308736]
[0.1313272 0.8119944] 1.12229710549 [-0.1683456  0.2469888]
[0.14816176 0.78729552] 1.11426403606 [-0.13467648  0.19759104]
[0.16162941 0.76753642] 1.10912412991 [-0.10774118  0.15807283]
[0.17240353 0.75172913] 1.10583559667 [-0.08619295  0.12645827]
[0.18102282 0.73908331] 1.10373174059 [-0.06895436  0.10116661]
[0.18791826 0.72896665] 1.102385917 [-0.05516348  0.08093329]
[0.19343461 0.72087332] 1.10152510532 [-0.04413079  0.06474664]
[0.19784768 0.71439865] 1.1009745981 [-0.03530463  0.0517973 ]
[0.20137815 0.70921892] 1.10062260339 [-0.0282437   0.04143785]
[0.20420252 0.70507514] 1.10039759063 [-0.02259496  0.03315028]
[0.20646201 0.70176011] 1.10025379358 [-0.01807598  0.02652022]
[0.20826961 0.69910809] 1.10016193234 [-0.01446078  0.02121617]
[0.20971569 0.69698647] 1.10010327627 [-0.01156862  0.01697294]
[0.21087255 0.69528918] 1

array([0.21549998, 0.68850003])

In [74]:
grad_descent2(f, gradf, array([0.5, 0.5]), 0.1)

[0.4431 0.5377] 1.1744661 [ 0.4552 -0.3016]
[0.39758 0.56786] 1.147646196 [ 0.36416 -0.24128]
[0.361164 0.591988] 1.13048391504 [ 0.291328 -0.193024]
[0.3320312 0.6112904] 1.11950202131 [ 0.2330624 -0.1544192]
[0.30872496 0.62673232] 1.11247518218 [ 0.18644992 -0.12353536]
[0.29007997 0.63908586] 1.10797926343 [ 0.14915994 -0.09882829]
[0.27516397 0.64896869] 1.105102882 [ 0.11932795 -0.07906263]
[0.26323118 0.65687495] 1.10326280324 [ 0.09546235 -0.06325011]
[0.25368494 0.66319996] 1.10208579712 [ 0.07636989 -0.05060008]
[0.24604796 0.66825997] 1.10133302857 [ 0.06109591 -0.04048007]
[0.23993836 0.67230797] 1.100851669 [ 0.04887673 -0.03238405]
[0.23505069 0.67554638] 1.10054392875 [ 0.03910138 -0.02590725]
[0.23114055 0.6781371 ] 1.10034723885 [ 0.03128111 -0.02072579]
[0.22801244 0.68020968] 1.10022156844 [ 0.02502488 -0.01658064]
[0.22550995 0.68186775] 1.10014130826 [ 0.02001991 -0.01326451]
[0.22350796 0.6831942 ] 1.10009007685 [ 0.01601592 -0.01061161]
[0.22190637 0.68425536] 1.

array([0.21550003, 0.68849998])

In [75]:
grad_descent2(f, gradf, array([0.99, 0.99]), 0.1)

[0.8351 0.9297] 1.5412213 [1.2392 0.4824]
[0.71118 0.88146] 1.382244084 [0.99136 0.38592]
[0.612044 0.842868] 1.28052621136 [0.793088 0.308736]
[0.5327352 0.8119944] 1.21544880935 [0.6344704 0.2469888]
[0.46928816 0.78729552] 1.17381690125 [0.50757632 0.19759104]
[0.41853053 0.76753642] 1.14718658341 [0.40606106 0.15807283]
[0.37792442 0.75172913] 1.13015446257 [0.32484884 0.12645826]
[0.34543954 0.73908331] 1.11926293164 [0.25987907 0.10116661]
[0.31945163 0.72896665] 1.11229957275 [0.20790326 0.08093329]
[0.2986613  0.72087332] 1.10784879968 [0.16632261 0.06474663]
[0.28202904 0.71439865] 1.10500492628 [0.13305808 0.0517973 ]
[0.26872324 0.70921892] 1.1031885445 [0.10644647 0.04143785]
[0.25807859 0.70507514] 1.10202901775 [0.08515717 0.03315027]
[0.24956287 0.70176011] 1.10128928684 [0.06812574 0.02652022]
[0.2427503  0.69910809] 1.10081775193 [0.05450059 0.02121618]
[0.23730024 0.69698647] 1.10051748392 [0.04360047 0.01697294]
[0.23294019 0.69528918] 1.10032652386 [0.03488039 0.013

array([0.21550003, 0.68850001])

Implement a two-dimensional version of the gradient descent algorithm to find optimal choices of a and b. (7 points)
Let the error function be L (loss function) and parameter    The goal is to find an optimal set of θ parameters to minimize L, as follows:

In this task, we do not need to care about the input and output. We can directly obtain loss through Request, so we only need to adjust the parameter gradient. Let the initial parameter be


As a result, I implemented the gradient descent funtion. It takes the f function, gradf funtion, and our inital guess of t, the step size alpha, and the EPS (Epsilon is a very small number to prevent any division by zero in the implementation.

Explain how you estimate the gradient given that you cannot directly compute the derivative (3 points)

Without knowing the exact function, we know the changing trend between parameters and loss, and we can calculate the gradient by calculating the derivative of loss and parameters (a, b). Loss decreases faster in the opposite direction of the gradient.

identify any numerical choices -- including but not limited to stopping criteria -- you made (3 points), and justify why you think they were reasonable choices (3 points).

In this task, when the number of iterations is less than the maximum 1000, the training is stopped. Another option is to check if the loss continues to decrease or the gradient changes near 0. The reasons for choosing these two methods are as follows: When the number of iterations is enough, the probability of obtaining the local or global minimum value of loss function by gradient descent method is increased and stops when the number of max iterations is reached. In the case of unfixed iteration steps, when iteration reaches the global or local minimum value, the gradient continues to fluctuate near the minimum value due to the existence of learning rate, and loss does not continue to decrease, so it stops.

It so happens that this error function has a local minimum and a global minimum. Find both locations (i.e. a, b values) querying the API as needed (5 points) and identify which corresponds to which (2 point). 

After training, we can find that the local minimum is 1.10000049995 when a = 0.2155, b = 0.6885; the global minimum is 1.00000149999 when a =0.7115, b = 0.1685.

1.Briefly discuss how you would have tested for local vs global minima if you had not known how many minima there were. (2 points)
If I had not known how many minima there were, I would try different corners and center, just like I did; I chose (0.99,0.01),(0.01,0.99)(0.01,0.01),(0.99,0.99), and (0.5,0.5) to test. Initializing multiple models with multiple different parameters is to starting the search from multiple different initial points, which may fall into different local minima, from which selection may obtain a result closer to the global minimum.
2.Simulated annealing accepts worse results than the current solution with some probability at each step, thus contributing to the jump out of the local minimum. In the iterative process of each step, the probability of accepting the "suboptimal solution" should gradually decrease with the passage of time, so as to ensure the stability of the algorithm.
3.Use random gradient descent. Different from the standard gradient descent method, the stochastic factor is added into the gradient calculation. Thus, even if it falls into the local minimum, its calculated gradient may not be zero, giving it a chance to jump out of the local minimum and continue the search.